crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,156 @@
+        state: State,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> SuiResult<Accumulator> {
+        if let Some(acc) = epoch_store.get_state_hash_for_checkpoint(&state.checkpoint_seq_num)? {
+            return Ok(acc);
+        }
+
+        let mut acc = Accumulator::default();
+
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.created
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        acc.remove_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.deleted
+                        .clone()
+                        .into_iter()
+                        .map(|obj_ref| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+        // TODO almost certainly not currectly handling "mutated" effects.
+        acc.insert_all(
+            state
+                .effects
+                .iter()
+                .flat_map(|fx| {
+                    fx.mutated
+                        .clone()
+                        .into_iter()
+                        .map(|(obj_ref, _)| obj_ref.2)
+                        .collect::<Vec<ObjectDigest>>()
+                })
+                .collect::<Vec<ObjectDigest>>(),
+        );
+
+        epoch_store.insert_state_hash_for_checkpoint(&state.checkpoint_seq_num, &acc)?;
+
+        epoch_store
+            .checkpoint_state_notify_read
+            .notify(&state.checkpoint_seq_num, &acc);
+
+        Ok(acc)
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
+    /// context. Can be called on non-consecutive epochs, e.g. to accumulate epoch 3 after
+    /// having last accumulated epoch 1.
+    pub fn accumulate_epoch(
+        &self,
+        epoch: &EpochId,
+        last_checkpoint_of_epoch: CheckpointSequenceNumber,
+        epoch_store: Arc<AuthorityPerEpochStore>,
+    ) -> Result<Accumulator, TypedStoreError> {
+        if let Some((_checkpoint, acc)) = self
+            .authority_store
+            .perpetual_tables
+            .root_state_hash_by_epoch
+            .get(epoch)?
+        {
+            return Ok(acc);
+        }
+
+        // Get the next checkpoint to accumulate (first checkpoint of the epoch)
+        // by adding 1 to the highest checkpoint of the previous epoch
+        let (epoch, (next_to_accumulate, mut root_state_hash)) = self
```

Comment:
Should we assert that this `epoch` is equal to `epoch argument - 1`?

Reply:
Makes sense

Reply:
I thought against this actually. I inteded for `accumulate_epoch` to be callable for abitrary epoch gaps. For example, if we need to accumulate from epoch 1 to 3. 

From a protocol perspective, we may be enforcing that each epoch has a root state hash, but that shouldn't be something that `StateAccumulator` worries about. If this invariant doesn't hold, `CheckpointBuilder` screwed up, and it should be the one enforcing it

Reply:
fair enough!